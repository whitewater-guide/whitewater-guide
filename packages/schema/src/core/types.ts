import { PageInfo } from '../__generated__/types';

/**
 * Relay-styles edge
 **/
export interface Edge<TNode> {
  node: TNode;
  cursor: CodegenCursor;
}

/**
 * Relay-styles connection
 **/
export interface RelayConnection<TNode> {
  edges: Array<Edge<TNode>>;
  pageInfo: PageInfo;
}
