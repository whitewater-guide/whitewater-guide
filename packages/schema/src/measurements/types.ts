export enum Unit {
  FLOW = 'flow',
  LEVEL = 'level',
}
