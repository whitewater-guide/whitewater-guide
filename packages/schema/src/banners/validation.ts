import { yupSchemas } from '@whitewater-guide/validation';
import * as yup from 'yup';

import {
  BannerInput,
  BannerKind,
  BannerPlacement,
  BannerSourceInput,
} from '../__generated__/types';

const BannerSourceSchema: yup.SchemaOf<BannerSourceInput> = yup
  .object({
    kind: yup.mixed().oneOf(Object.values(BannerKind)),
    url: yup.string().required(),
  })
  .noUnknown();

export const BannerInputSchema: yup.SchemaOf<BannerInput> = yup
  .object({
    id: yup.string().uuid().nullable(),
    slug: yup.string().slug().required(),
    name: yup.string().required(),
    priority: yup.number().integer().required(),
    enabled: yup.bool().required(),
    placement: yup.mixed().oneOf(Object.values(BannerPlacement)),
    source: BannerSourceSchema.clone(),
    link: yup.string().url().https().nullable(true),
    extras: yup.object().nullable(true),
    regions: yup.array(yupSchemas.refInput()),
    groups: yup.array(yupSchemas.refInput()),
  })
  .strict(true)
  .noUnknown(true);
