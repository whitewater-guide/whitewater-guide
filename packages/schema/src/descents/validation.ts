import * as yup from 'yup';

import { DescentInput, DescentLevelInput } from '../__generated__/types';

export const DescentLevelInputSchema: yup.SchemaOf<DescentLevelInput> = yup
  .object({
    unit: yup.string().max(20).optional().nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema: yup.SchemaOf<DescentInput> = yup
  .object({
    id: yup.string().uuid().optional().nullable(),
    sectionId: yup.string().uuid().required().nullable(false),
    // Something is wrong with date type
    // https://github.com/jquense/yup/issues/1243
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    startedAt: yup.date().required() as any,
    duration: yup.number().integer().min(0).notRequired().nullable(),
    level: DescentLevelInputSchema.clone().notRequired().nullable(),
    comment: yup.string().notRequired().nullable(),
    public: yup.bool().notRequired().nullable(),
  })
  .strict(true)
  .noUnknown();
