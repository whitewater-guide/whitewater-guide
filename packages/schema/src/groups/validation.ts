import * as yup from 'yup';

import { GroupInput } from '../__generated__/types';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema: yup.SchemaOf<GroupInput> = yup
  .object({
    id: yup.string().uuid().defined().nullable(),
    name: yup.string().nonEmpty(),
    sku: yup
      .string()
      .defined()
      .max(255)
      .matches(GROUP_SKU, 'yup:string.sku')
      .nullable(),
  })
  .strict(true)
  .noUnknown();
