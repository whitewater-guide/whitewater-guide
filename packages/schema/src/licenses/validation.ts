import * as yup from 'yup';

import { LicenseInput } from '../__generated__/types';

export const LicenseInputSchema: yup.SchemaOf<LicenseInput> = yup
  .object({
    url: yup.string().url().nullable(true),
    name: yup.string().min(2).defined().nullable(false),
    slug: yup.string().slug().nullable(true),
  })
  .strict(true)
  .noUnknown();
