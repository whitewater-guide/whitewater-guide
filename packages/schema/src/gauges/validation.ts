import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { mixed, object, string } from 'yup';

import type { GaugeInput } from '../__generated__/types';
import { PointInputSchema } from '../points';

export const GaugeInputSchema: ObjectSchema<GaugeInput> = object({
  id: string().uuid().nullable(),
  name: string().nonEmpty(),
  code: string().nonEmpty(),
  levelUnit: string().nullable(),
  flowUnit: string().nullable(),
  timezone: string().nullable(),
  location: PointInputSchema.clone().nullable(),
  requestParams: mixed().nullable(),
  url: string().url().nullable(),
  source: yupSchemas.refInput().required(),
})
  .strict(true)
  .noUnknown();
