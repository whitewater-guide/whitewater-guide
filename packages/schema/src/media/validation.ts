import * as yup from 'yup';

import { MediaInput, MediaKind } from '../__generated__/types';
import { LicenseInputSchema } from '../licenses';

export const MediaInputSchema: yup.SchemaOf<MediaInput> = yup
  .object({
    id: yup.string().uuid().nullable(),
    description: yup.string().nullable(),
    copyright: yup.string().nullable(),
    license: LicenseInputSchema.clone().nullable(true),
    url: yup.string().nonEmpty(),
    kind: yup.mixed().defined().oneOf(Object.values(MediaKind)),
    resolution: yup
      .array()
      .min(2)
      .max(2)
      .of(yup.number().required().integer().positive())
      .optional()
      .nullable(),
    weight: yup.number().integer().nullable(),
  })
  .strict(true)
  .noUnknown();
