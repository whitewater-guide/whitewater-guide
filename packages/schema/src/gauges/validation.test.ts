import { createSafeValidator } from '@whitewater-guide/validation';

import { GaugeInput } from '../__generated__/types';
import { GaugeInputSchema } from './validation';

const validator = createSafeValidator(GaugeInputSchema);

type TestValue = [string, GaugeInput];
type IncorrectTestValue = [string, any];

const required: GaugeInput = {
  name: 'Gauge',
  code: 'g001',
  source: {
    id: '7df5722e-2d30-4993-890b-c062de184f78',
  },
};

const gauge: GaugeInput = {
  ...required,
  id: '712fcc21-6d0b-475a-bf53-2cb1892767fc',
  levelUnit: 'm',
  flowUnit: 'm3/s',
  location: {
    id: '28de619b-a05b-430d-b835-527183b4f1dd',
    name: 'Gauge g001',
    description: null,
    coordinates: [-170, 80, 123.5],
    kind: 'gauge',
  },
  requestParams: { foo: 'bar' },
  url: 'http://google.com',
  timezone: 'Europe/Moscow',
};

const correctValues: TestValue[] = [
  ['full value', gauge],
  ['only required fields', required],
  [
    'nulled value',
    {
      ...gauge,
      id: null,
      levelUnit: null,
      flowUnit: null,
      location: null,
      requestParams: null,
      url: null,
      timezone: null,
    },
  ],
  [
    'value with empty string url',
    {
      ...gauge,
      url: '',
    },
  ],
];

const incorrectValues: IncorrectTestValue[] = [
  [
    'all',
    {
      id: 'aaaa',
      name: '',
      code: '',
      levelUnit: '',
      flowUnit: '',
      location: {
        id: 'zzz',
        name: 'z',
        description: null,
        coordinates: [-190, 800, -123.5],
        kind: 'foo',
      },
      requestParams: {},
      url: 'http://google.',
      source: {
        id: 'bar',
      },
    },
  ],
  ['undefined name', { ...gauge, name: undefined }],
  ['undefined code', { ...gauge, code: undefined }],
  ['undefined source', { ...gauge, source: undefined }],
  ['null source', { ...gauge, source: null }],
  ['extra field', { ...gauge, foo: 'bar' }],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  const error = validator(value);
  expect(error).not.toBeNull();
  expect(error).toMatchSnapshot();
});
