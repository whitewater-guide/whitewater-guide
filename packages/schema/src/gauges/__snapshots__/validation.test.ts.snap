// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for all 1`] = `
Object {
  "code": "yup:string.nonEmpty",
  "id": Object {
    "key": "yup:string.uuid",
    "options": Object {
      "path": "id",
    },
  },
  "location": Object {
    "coordinates": Array [
      "yup:number.longitude",
      "yup:number.latitude",
    ],
    "id": Object {
      "key": "yup:string.uuid",
      "options": Object {
        "path": "location.id",
      },
    },
    "kind": Object {
      "key": "yup:mixed.oneOf",
      "options": Object {
        "values": "put-in, put-in-alt, put-in-road, take-out, take-out-alt, take-out-road, waterfall, rapid, portage, playspot, hazard, river-campsite, wild-camping, paid-camping, gauge, hike-waypoint, bridge, dam, power-plant, kayak-shop, other",
      },
    },
  },
  "name": "yup:string.nonEmpty",
  "source": Object {
    "id": Object {
      "key": "yup:string.uuid",
      "options": Object {
        "path": "source.id",
      },
    },
  },
  "url": Object {
    "key": "yup:string.url",
    "options": Object {
      "path": "url",
    },
  },
}
`;

exports[`should be invalid for extra field 1`] = `
Object {
  "": Object {
    "key": "yup:object.noUnknown",
    "options": Object {
      "path": "this",
    },
  },
}
`;

exports[`should be invalid for null source 1`] = `
Object {
  "source": Object {
    "key": "yup:mixed.notType",
    "options": Object {
      "type": "object",
    },
  },
}
`;

exports[`should be invalid for undefined code 1`] = `
Object {
  "code": "yup:string.nonEmpty",
}
`;

exports[`should be invalid for undefined name 1`] = `
Object {
  "name": "yup:string.nonEmpty",
}
`;

exports[`should be invalid for undefined source 1`] = `
Object {
  "source": "yup:mixed.defined",
}
`;
